component/commons/handler/src/main/java/org/wso2/carbon/identity/password/expiry/validation/handler/PasswordExpiryValidationHandler.java

```diff
@@ -0,0 +1,179 @@
+ */
+public class PasswordExpiryValidationHandler extends AbstractEventHandler {
+
+    private static final Log log = LogFactory.getLog(PasswordExpiryValidationHandler.class);
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        String username = (String) event.getEventProperties().get(IdentityEventConstants.EventProperty.USER_NAME);
+        String tenantDomain = (String) event.getEventProperties()
+                .get(IdentityEventConstants.EventProperty.TENANT_DOMAIN);
+        UserStoreManager userStoreManager = (UserStoreManager) event.getEventProperties()
+                .get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
+        String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+
+        try {
+            if (isPasswordExpired(tenantDomain, tenantAwareUsername, userStoreManager)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("User: " + username + " password is expired.");
+                }
+                throw new IdentityEventException(PasswordExpiryValidationConstants.PASSWORD_EXPIRED_ERROR_MESSAGE);
+            }
+        } catch (UserStoreException e) {
+            throw new IdentityEventException("UserStore Exception occurred while password expiry validation", e);
+        }
+    }
+     /**
+     * Check whether the password is expired or not.
+     * @param tenantDomain user tenant domain
+     * @param tenantAwareUsername The tenant aware username of the user trying to authenticate
+     * @param userStoreManager
+     * @return true if password is expired or password last update time is null
+     * @throws UserStoreException
+     * @throws ParseException
+     */
+    private boolean isPasswordExpired(String tenantDomain, String tenantAwareUsername,
+                                      UserStoreManager userStoreManager) throws UserStoreException {
+
+        String passwordLastChangedTime = getPasswordLastChangeTime(tenantDomain, tenantAwareUsername, userStoreManager);
+        int passwordExpiryInDays =  getPasswordExpiryInDaysConfig(tenantDomain);
+
+        long passwordChangedTime = 0;
+        if (passwordLastChangedTime != null) {
+            passwordChangedTime = Long.parseLong(passwordLastChangedTime);
+        }
+        int daysDifference = 0;
+        long currentTimeMillis = System.currentTimeMillis();
+        if (passwordChangedTime > 0) {
+            Calendar currentTime = Calendar.getInstance();
+            currentTime.add(Calendar.DATE, (int) currentTime.getTimeInMillis());
+            daysDifference = (int) ((currentTimeMillis - passwordChangedTime) / (1000 * 60 * 60 * 24));
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("User: " + tenantAwareUsername + " password is updated before " + daysDifference + " Days");
+        }
+        return (daysDifference > passwordExpiryInDays || passwordLastChangedTime == null);
```

Comment:
If the password has never been changed should we ask them to change password ? Why is that ? Are we setting something to LastChangedTime when initially a user is onboarded ?

Reply:
As per the current implementation of IS connector[1] for password expiry, the lastPasswordUpdateTime claim is updated when the user is created. This is also done by an event handler. This new event handler is also shipped with that connector.

[1] - https://store.wso2.com/store/assets/isconnector/details/502efeb1-cc59-4b62-a197-8c612797933c

[2] - https://github.com/wso2-extensions/identity-outbound-auth-passwordPolicy/blob/master/docs/config.md